runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowRunner.java

```diff
@@ -550,16 +554,34 @@ public DataflowPipelineJob run(Pipeline pipeline) {
       hooks.modifyEnvironmentBeforeSubmission(newJob.getEnvironment());
     }
 
-    if (!isNullOrEmpty(options.getDataflowJobFile())) {
-      try (PrintWriter printWriter = new PrintWriter(
-          new File(options.getDataflowJobFile()))) {
-        String workSpecJson = DataflowPipelineTranslator.jobToString(newJob);
+    if (!isNullOrEmpty(options.getDataflowJobFile())
+        || !isNullOrEmpty(options.getTemplateLocation())) {
+      boolean isTemplate = !isNullOrEmpty(options.getTemplateLocation());
+      if (isTemplate) {
+        checkArgument(isNullOrEmpty(options.getDataflowJobFile()),
+            "--dataflowJobFile and --templateLocation are mutually exclusive.");
+      }
+      String fileLocation = firstNonNull(
+          options.getTemplateLocation(), options.getDataflowJobFile());
+      String workSpecJson = DataflowPipelineTranslator.jobToString(newJob);
+      try {
+        WritableByteChannel writer =
+            IOChannelUtils.create(fileLocation, MimeTypes.TEXT);
+        PrintWriter printWriter = new PrintWriter(Channels.newOutputStream(writer));
         printWriter.print(workSpecJson);
```

Comment:
with this syntax, you need the `.close()` still on the `printWriter`. You only get AutoCloseable magic using the [try-with-resources syntax](https://docs.oracle.com/javase/tutorial/essential/exceptions/tryResourceClose.html):

```java
try (create some autocloseable) {
   do stuff
} // close is automatically called here.
```